tsa/client.go

```diff
@@ -0,0 +1,551 @@
+	baggageclaimForwardAddr = "0.0.0.0:7788"
+)
+
+// Client is used to communicate with a pool of remote SSH gateways.
+type Client struct {
+	Hosts    []string
+	HostKeys []ssh.PublicKey
+
+	PrivateKey *rsa.PrivateKey
+
+	Worker atc.Worker
+}
+
+// RegisterOptions contains required configuration for the registration.
+type RegisterOptions struct {
+	// The local Garden network and address to forward through the SSH gateway.
+	LocalGardenNetwork string
+	LocalGardenAddr    string
+
+	// The local Baggageclaim network and address to forward through the SSH
+	// gateway.
+	LocalBaggageclaimNetwork string
+	LocalBaggageclaimAddr    string
+
+	// Under normal circumstances, the connection is kept alive by continuously
+	// sending a keepalive request to the SSH gateway. When the context is
+	// canceled, the keepalive loop is stopped, and the connection will break
+	// after it has been idle for this duration, if configured.
+	DrainTimeout time.Duration
+
+	// RegisteredFunc is called when the initial registration has completed.
+	//
+	// The function must be careful not to take too long or become deadlocked, or
+	// else the SSH connection can starve.
+	RegisteredFunc func()
+
+	// HeartbeatedFunc is called on each heartbeat after registration.
+	//
+	// The function must be careful not to take too long or become deadlocked, or
+	// else the SSH connection can starve.
+	HeartbeatedFunc func()
+}
+
+// Register invokes the 'forward-worker' command, proxying traffic through the
+// tunnel and to the configured Garden/Baggageclaim addresses. It will also
+// continuously keep the connection alive. The SSH gateway will continuously
+// heartbeat the worker.
+//
+// If the context is canceled, heartbeating is immediately stopped and the
+// remote SSH gateway will wait for connections to drain. If a DrainTimeout is
+// configured, the connection will be terminated after no data has gone to/from
+// the SSH gateway for the configured duration.
+func (client *Client) Register(ctx context.Context, opts RegisterOptions) error {
+	logger := lagerctx.FromContext(ctx)
+
+	sshClient, tcpConn, err := client.dial(ctx, opts.DrainTimeout)
+	if err != nil {
+		logger.Error("failed-to-dial", err)
+		return err
+	}
+
+	defer sshClient.Close()
+
+	go client.keepAlive(ctx, sshClient, tcpConn)
+
+	gardenListener, err := sshClient.Listen("tcp", gardenForwardAddr)
+	if err != nil {
+		logger.Error("failed-to-listen-for-garden", err)
+		return err
+	}
+
+	go proxyListenerTo(ctx, gardenListener, opts.LocalGardenNetwork, opts.LocalGardenAddr)
+
+	baggageclaimListener, err := sshClient.Listen("tcp", baggageclaimForwardAddr)
+	if err != nil {
+		logger.Error("failed-to-listen-for-baggageclaim", err)
+		return err
+	}
+
+	go proxyListenerTo(ctx, baggageclaimListener, opts.LocalBaggageclaimNetwork, opts.LocalBaggageclaimAddr)
+
+	eventsR, eventsW := io.Pipe()
+	defer eventsW.Close()
+
+	events := NewEventReader(eventsR)
+	go func() {
+		for {
```

Comment:
Should this for also be watching the context for cancellations? Eventually `events.Next()` will return an `io.EOF`, but we could still be receiving and handling events after the context passed to `Register()` has been cancelled.

Reply:
This shouldn't watch the context; if events are being sent, something has to be reading them or else the SSH connection may hang (on the Go side). Closing the context is also used as part of the TSA SSH protocol (it sends a signal), which could theoretically trigger subsequent events which may need to be read. Reading until EOF should be enough; it's guaranteed to happen because of the `defer sshClient.Close()` and `defer eventsW.Close()`.